File: Java/Medium/MemoryLimitExceeded/count-the-number-of-square-free-subsets_0.java
class Solution {
    public int squareFreeSubsets(int[] nums) {
        
        int n = nums.length;
        int mod = 1000000007;
        int[] count = new int[31];
        for (int i = 0; i < n; i++) {
            count[nums[i]]++;
        }
        int[] dp = new int[1 << 30];
        dp[0] = 1;
        for (int i = 1; i <= 30; i++) {
            int[] tmp = new int[1 << 30];
            for (int j = 0; j < 1 << 30; j++) {
                if (dp[j] == 0) {
                    continue;
                }
                for (int k = 1; k <= count[i]; k++) {
                    int mask = j;
                    for (int l = 0; l < k; l++) {
                        mask |= 1 << (i * l);
                    }
                    tmp[mask] += dp[j];
                    if (tmp[mask] >= mod) {
                        tmp[mask] -= mod;
                    }
                }
            }
            for (int j = 0; j < 1 << 30; j++) {
                dp[j] += tmp[j];
                if (dp[j] >= mod) {
                    dp[j] -= mod;
                }
            }
        }
        int ans = 0;
        for (int i = 0; i < 1 << 30; i++) {
            int cnt = 0;
            for (int j = 0; j < 30; j++) {
                if ((i & (1 << j)) != 0) {
                    cnt++;
                }
            }
            if (cnt % 2 == 0) {
                ans += dp[i];
                if (ans >= mod) {
                    ans -= mod;
                }
            } else {
                ans -= dp[i];
                if (ans < 0) {
                    ans += mod;
                }
            }
        }
        return ans;
    }
}

File: C++/Medium/TimeLimitExceeded/count-the-number-of-square-free-subsets_0.cpp
class Solution {
public:
    int squareFreeSubsets(vector<int>& nums) {
        
        int n = nums.size();
        int mod = 1e9 + 7;
        vector<int> cnt(31, 0);
        for (int i = 0; i < n; i++) {
            cnt[nums[i]]++;
        }
        vector<int> dp(1 << 30, 0);
        dp[0] = 1;
        for (int i = 1; i <= 30; i++) {
            for (int j = (1 << 30) - 1; j >= 0; j--) {
                for (int k = 1; k <= cnt[i]; k++) {
                    int mask = (1 << i) - 1;
                    if ((j & mask) == 0) {
                        dp[j | mask] = (dp[j | mask] + dp[j]) % mod;
                    }
                    mask <<= i;
                }
            }
        }
        int ans = 0;
        for (int i = 0; i < (1 << 30); i++) {
            if (__builtin_popcount(i) & 1) {
                ans = (ans - dp[i] + mod) % mod;
            } else {
                ans = (ans + dp[i]) % mod;
            }
        }
        return ans;
    }
};